libs/sdk-core/src/greenlight/node_api.rs

```diff
@@ -810,6 +811,69 @@ impl NodeAPI for Greenlight {
     fn legacy_derive_bip32_key(&self, path: Vec<ChildNumber>) -> Result<ExtendedPrivKey> {
         Self::legacy_derive_bip32_key(self.sdk_config.network, &self.signer, path)
     }
+
+    async fn stream_custom_messages(&self) -> Result<mpsc::Receiver<CustomMessage>> {
+        let mut stream = self
+            .get_client()
+            .await?
+            .stream_custommsg(gl_client::pb::StreamCustommsgRequest {})
+            .await?
+            .into_inner();
+        let (tx, rx) = mpsc::channel(100);
+
+        // Copy the greenlight stream to the mpsc::channel to map to our
+        // internal CustomMessage struct.
+        tokio::spawn(async move {
+            loop {
+                let msg = match stream.message().await {
+                    Ok(m) => match m {
+                        Some(m) => m,
+                        None => return,
+                    },
+                    Err(_) => return,
```

Comment:
I think the way we fetch messages in a reliable way from the custom message stream could be similar to the way we do it for the invoice stream (stream_incoming). We have a task that wraps the call to the stream_incoming and whenever there is an error it sleeps some delay and then create the stream again (track_invoices).

Reply:
Done. I had in mind to put that retry logic in the caller. But it indeed makes more sense here.

Reply:
Looks good now. 
The only issue I see now is regarding the tokio task that is running here and is not controlled in a sense that there is no code that stops it.
Do we really need a task here? (and channel to communicate?).
Since the caller (transport) already running in a task and call this method from within the task it would look cleaner if this method returns a stream of custom message by transforming the gRpc stream without background processing.
This way the caller is responsible for running it in the background (or not) and we don't need to worry about managing this task.
What do you think?

Reply:
For some reason I totally misplaced this comment, and thought it was about `transport.rs`. Anyway, good points.

I think it would be great to be able to map that stream into another stream while reading it. Like `stream.map(..)`, or some kind of Generator, yield return kind of pattern. I couldn't find any though.

It is sort-of stopped, in the sense that when the receiver drops, the method returns, so the task is completed. Also, when the sender drops (silently return), the receiver will get an empty message, therefore it knows the stream disconnected. There could be useful info in the error, it should probably be logged. 

I couldn't find any good way to map the one stream to the other without usnig the task, or blocking the caller, which I don't like. Also, you need to return the receiver, so you can't infinitely loop here without returning something?

Reply:
tonic streaming implements tokio::Stream https://github.com/hyperium/tonic/blob/master/tonic/src/codec/decode.rs#L384

So perhaps we can use tokio stream mapping to return the steam with our own model? https://tokio.rs/tokio/tutorial/streams ,(see adapters section)

Reply:
Updated to map the stream with `tokio_stream`. Can you check again?

Reply:
That looks great!